### PR TITLE
Update 1.20.2 & Gradle 8.3, & Loom updates

### PR DESCRIPTION
**Pull request by sakura-ryoko**
Hey awesome MOD, so I wanted to learn to port stuff, this one was simple enough for me to do.

### CODE REVIEW

**Comment by sakura-ryoko** [comment]
pls forgive the mess, but it should build and install ok... maybe make sure it works with the updated LP 1.20.2 releases.

**Comment by MattiDragon** [comment]
I'll do a proper review tomorrow, but with a quick look I can see some issues:
- Why did you move/add assets to `bin`? That's not standard for fabric projects
- Why did you change my buildscript (besides the archives base name moving)? I changed from the default dependencies in the example mod for a reason, mine are easier to read for experienced devs.
- Why does the mod icon show as changed?
- Why is there a commented out import? Just remove unused ones

File: bin/main/fabric.mod.json
Comment (MattiDragon) [inline bin/main/fabric.mod.json:1]: Why was this file added? If this is something that your IDE generated then it should be in the gitignore, if you added it yourself then it should be removed

File: bin/main/universal_perms.mixins.json
Comment (MattiDragon) [inline bin/main/universal_perms.mixins.json:1]: Why was this file added? If this is something that your IDE generated then it should be in the gitignore, if you added it yourself then it should be removed

File: bin/main/universal_perms_icon.png
Comment (MattiDragon) [inline bin/main/universal_perms_icon.png:1]: Why was this file added? If this is something that your IDE generated then it should be in the gitignore, if you added it yourself then it should be removed

File: bin/main/universal_perms_icon.xcf
Comment (MattiDragon) [inline bin/main/universal_perms_icon.xcf:1]: Why was this file added? If this is something that your IDE generated then it should be in the gitignore, if you added it yourself then it should be removed

File: build.gradle [hunk 1]
```diff
@@ -1,27 +1,45 @@
...
 
+loom {
+    splitEnvironmentSourceSets()
+
+        mods {
+                "modid" {
+                        sourceSet sourceSets.main
+                        sourceSet sourceSets.client
+                }
+        }
```
Comment (MattiDragon) [inline build.gradle:26]: Unnessecary and uses the wrong id, please remove

File: build.gradle [hunk 2]
```diff
@@ -1,27 +1,45 @@
...
 dependencies {
-    minecraft "com.mojang:minecraft:$minecraft_version"
-    mappings "net.fabricmc:yarn:$yarn_mappings:v2"
-    modImplementation "net.fabricmc:fabric-loader:$loader_version"
+        // To change the versions see the gradle.properties file
+        minecraft "com.mojang:minecraft:${project.minecraft_version}"
+        mappings "net.fabricmc:yarn:${project.yarn_mappings}:v2"
+        modImplementation "net.fabricmc:fabric-loader:${project.loader_version}"
+        // Fabric API. This is technically optional, but you probably want it anyway.
+        modImplementation "net.fabricmc.fabric-api:fabric-api:${project.fabric_version}"
```
Comment (MattiDragon) [inline build.gradle:36]: Please don't add back the comments that I deleted from the example mod, same for the unnessecary braces and `project.`.

File: build.gradle [hunk 3]
```diff
@@ -35,26 +53,29 @@ processResources {
...
+                }
+        }
+        repositories {
         }
-    }
-
-    repositories {
 
-    }
 }
```
Comment (MattiDragon) [inline build.gradle:81]: Why was this changed?

File: src/main/java/io/github/mattidragon/universalperms/UniversalPerms.java [hunk 4]
```diff
@@ -1,5 +1,13 @@
 package io.github.mattidragon.universalperms;
 
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+// UNUSED: import java.util.function.Predicate;
```
Comment (MattiDragon) [inline src/main/java/io/github/mattidragon/universalperms/UniversalPerms.java:8]: If an import is unused, just remove it. I'd also prefer if you didn't change the imports from the default order intellij puts them in

File: src/main/java/io/github/mattidragon/universalperms/UniversalPerms.java [hunk 5]
```diff
@@ -10,40 +18,43 @@
...
-            alterNode(dispatcher.getRoot(), new ArrayDeque<>(), new HashMap<>());
-            LOGGER.info("Applied cursed permissions!");
-        });
+        CommandRegistrationCallback.EVENT.addPhaseOrdering(Event.DEFAULT_PHASE,
+                new Identifier("universal_perms", "after"));
+        CommandRegistrationCallback.EVENT.register(new Identifier("universal_perms", "after"),
+                (dispatcher, registryAccess, environment) -> {
+                    alterNode(dispatcher.getRoot(), new ArrayDeque<>(), new HashMap<>());
+                    LOGGER.info("Applied cursed permissions!");
+                });
```
Comment (MattiDragon) [inline src/main/java/io/github/mattidragon/universalperms/UniversalPerms.java:33]: Please don't unnessecarily reformat my code

File: src/main/java/io/github/mattidragon/universalperms/UniversalPerms.java [hunk 6]
```diff
@@ -10,40 +18,43 @@
...
     }
 
-    private static void alterNode(CommandNode<ServerCommandSource> node, Deque<String> location, Map<CommandNode<ServerCommandSource>, String> visited) {
+    private static void alterNode(CommandNode<ServerCommandSource> node, Deque<String> location,
+            Map<CommandNode<ServerCommandSource>, String> visited) {
```
Comment (MattiDragon) [inline src/main/java/io/github/mattidragon/universalperms/UniversalPerms.java:37]: Again, please don't reformat for no reason

File: src/main/java/io/github/mattidragon/universalperms/UniversalPerms.java [hunk 7]
```diff
@@ -10,40 +18,43 @@
...
         var name = node.getName();
         if (!name.isEmpty())
             location.addLast(name);
 
         var nodeLocation = String.join(".", location);
 
         if (visited.containsKey(node)) {
-            LOGGER.warn("Encountered same node a two different places, '{}' and '{}'. This shouldn't happen!", visited.get(node), nodeLocation);
+            LOGGER.warn("Encountered same node a two different places, '{}' and '{}'. This shouldn't happen!",
+                    visited.get(node), nodeLocation);
```
Comment (MattiDragon) [inline src/main/java/io/github/mattidragon/universalperms/UniversalPerms.java:46]: Again, please don't reformat for no reason

File: src/main/java/io/github/mattidragon/universalperms/UniversalPerms.java [hunk 8]
```diff
@@ -10,40 +18,43 @@
...
             return;
         }
 
         var permission = createPermission("use", location);
         var requirement = node.getRequirement();
-        ((CommandNodeAccess)node).setRequirement((ServerCommandSource source) -> (source.getEntity() == null ? TriState.DEFAULT : Permissions.getPermissionValue(source, permission)).orElseGet(() -> requirement.test(source)));
+        ((CommandNodeAccess) node)
+                .setRequirement((ServerCommandSource source) -> (source.getEntity() == null ? TriState.DEFAULT
+                        : Permissions.getPermissionValue(source, permission))
+                        .orElseGet(() -> requirement.test(source)));
```
Comment (MattiDragon) [inline src/main/java/io/github/mattidragon/universalperms/UniversalPerms.java:57]: This should probably have been reformatted, but you did it in a way that's really confusing to read

File: src/main/java/io/github/mattidragon/universalperms/mixin/ServerPlayerEntityMixin.java [hunk 9]
```diff
@@ -6,11 +6,11 @@
 import io.github.mattidragon.universalperms.UniversalPerms;
 import me.lucko.fabric.api.permissions.v0.Options;
 import net.minecraft.entity.player.PlayerEntity;
-import net.minecraft.network.encryption.PlayerPublicKey;
+// FIXME: import net.minecraft.network.encryption.PlayerPublicKey;
```
Comment (MattiDragon) [inline src/main/java/io/github/mattidragon/universalperms/mixin/ServerPlayerEntityMixin.java:9]: Just remove unused imports

File: src/main/java/io/github/mattidragon/universalperms/mixin/ServerPlayerEntityMixin.java [hunk 10]
```diff
@@ -6,11 +6,11 @@
...
 import net.minecraft.server.network.ServerPlayerEntity;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
-import org.jetbrains.annotations.Nullable;
+// FIXME: import org.jetbrains.annotations.Nullable;
```
Comment (MattiDragon) [inline src/main/java/io/github/mattidragon/universalperms/mixin/ServerPlayerEntityMixin.java:13]: Just remove unused imports

File: src/main/java/io/github/mattidragon/universalperms/mixin/ServerPlayerEntityMixin.java [hunk 11]
```diff
@@ -20,8 +20,12 @@ public abstract class ServerPlayerEntityMixin extends PlayerEntity {
     @Unique
     private boolean universal_perms$is_checking_permission;
 
-    public ServerPlayerEntityMixin(World world, BlockPos pos, float yaw, GameProfile gameProfile, @Nullable PlayerPublicKey publicKey) {
-        super(world, pos, yaw, gameProfile, publicKey);
+    // API call fails with publicKey --> Seems to have been removed from API
+    // public ServerPlayerEntityMixin(World world, BlockPos pos, float yaw,
+    // GameProfile gameProfile, @Nullable PlayerPublicKey publicKey) {
```
Comment (MattiDragon) [inline src/main/java/io/github/mattidragon/universalperms/mixin/ServerPlayerEntityMixin.java:25]: Unnesssary comment. If the method changed, just update the mixin to reflect that, don't leave the old code behind

File: src/main/resources/fabric.mod.json [hunk 12]
```diff
@@ -23,9 +23,9 @@
     "universal_perms.mixins.json"
   ],
   "depends": {
-    "fabricloader": ">=0.14.9",
+    "fabricloader": "0.14.22",
     "fabric": "*",
-    "minecraft": ">=1.19.2",
+    "minecraft": "1.20.2",
```
Comment (MattiDragon) [inline src/main/resources/fabric.mod.json:28]: You removed the `>=` from the versions. This is bad especially for the loader version as any loader version over the one we use should work. This mod is also pretty mc version agnostic so it should also stay on the mc version

**Review by MattiDragon — CHANGES_REQUESTED** [review]

**Comment by MattiDragon** [comment]
I'll probably port this myself once I finish with other mods (which might take a while), but I'll let you try to get your code up to standard and I'll merge this if you fix the issues

**Comment by sakura-ryoko** [comment]
Its all good, I was just messing with coding for the first time.  You can simply reject the merge if you wish & I'll delete the fork.
Also, most of the things you've highlighted are .gitignore and IDE (in my case, VSCode) related items, and the build script is a learning curve issue.

I am much more comfortable letting you make the changes.

**Pull request closed without merging**